### PR TITLE
Use required fields in classes and exceptions

### LINKED ISSUE #4014: Improve mapping for classes and exceptions with required properties
The C# mapping for a simple class Person:
```slice
class Person {
    name: string
    age: int32
    tag(1) phoneNumber: string?
}
```

is currently:

```csharp
public partial class Person : SliceClass
{
    public string Name { get; set; }

    public int Age { get; set; }

    public string? PhoneNumber { get; set; }

    /// <summary>Constructs a new instance of <see cref="Person" />.</summary>
    public Person(
        string name,
        int age,
        string? phoneNumber)
    {
        this.Name = name;
        this.Age = age;
        this.PhoneNumber = phoneNumber;
    }

    /// <summary>Constructs a new instance of <see cref="Person" />.</summary>
    public Person(
        string name,
        int age)
    {
        this.Name = name;
        this.Age = age;
    }

    /// <summary>Constructs a new instance of <see cref="Person"/> for the Slice decoder.</summary>
    [global::System.ComponentModel.EditorBrowsable(global::System.ComponentModel.EditorBrowsableState.Never)]
    public Person()
    {
        this.Name = default!;
    }

    // encode/decode logic
}
```

We provide:
 - a primary constructor to initialize all properties (including nullable properties and value types that have a reasonable default)
 - (optionally) another constructor to initialize all non-tagged properties
 - a hidden parameterless constructor used for decoding. This paramerless ctor assigns `default!` to all properties with a non-nullable reference type.

This structure is dictated in large part by the Slice1 decoding logic for classes and exceptions, where we decode the fields of the most derived slice first and then go up the inheritance chain.

## Proposal:

 - Always make the parameterless constructor public / non-hidden, with an empty body (no more `= default!`)
 - Add `required` to all properties with a non-nullable reference type
 - When there is at least one required property, add `[SetsRequiredMembers]` to the generated constructor(s) other than the parameterless constructor.

This gives users more flexibility while making the generated code cleaner.

### PR DESCRIPTION
**Pull request by bernardnormier**
This PR updates the C# mapping for Slice classes and exceptions to use 'required' fields for non-nullable reference types.

Fixes #4014.

### CODE REVIEW

File: tools/slicec-cs/src/generators/exception_generator.rs [hunk 1]
```diff
@@ -49,23 +49,21 @@ pub fn generate_exception(exception_def: &Exception) -> CodeBlock {
...
-                    ),
+                    format!(r#"Constructs a new instance of <see cref="{}"/>."#, &exception_name),
                 )
-                .set_body(initialize_required_fields(fields))
                 .build(),
         );
+
+        exception_class_builder.add_block(one_shot_constructor(exception_def));
+
+        // TODO: generate secondary constructor like for classes.
```
Comment (bernardnormier) [inline tools/slicec-cs/src/generators/exception_generator.rs:64]: It's unfortunate that class_generator and exception_generator are so different, when the logic and generated code are almost identical. Fixing this inconsistency is outside the scope of this PR.

File: tools/slicec-cs/src/generators/class_generator.rs [hunk 2]
```diff
@@ -70,41 +70,39 @@ pub fn generate_class(class_def: &Class) -> CodeBlock {
...
-        // The constructor needs to be public for System.Activator.CreateInstance.
-        let mut decode_constructor = FunctionBuilder::new("public", "", &class_name, FunctionType::BlockBody);
-        decode_constructor.add_never_editor_browsable_attribute();
-        decode_constructor.add_comment(
-            "summary",
-            format!(r#"Constructs a new instance of <see cref="{class_name}"/> for the Slice decoder."#),
-        );
-        decode_constructor.set_body(initialize_required_fields(&fields));
-        class_builder.add_block(decode_constructor.build());
+        // Secondary constructor for all fields minus those with optional types.
```
Comment (bernardnormier) [inline tools/slicec-cs/src/generators/class_generator.rs:90]: The comment should use "nullable" for consistency with terminology in the code below.

File: tools/slicec-cs/src/generators/class_generator.rs [hunk 3]
```diff
@@ -70,41 +70,39 @@ pub fn generate_class(class_def: &Class) -> CodeBlock {
...
+        // This constructor is only generated if necessary
+        if (non_nullable_fields.len() + non_nullable_base_fields.len() > 0)
```
Comment (bernardnormier) [inline tools/slicec-cs/src/generators/class_generator.rs:92]: Is there a more elegant way to write this expression?

**Review by externl — APPROVED** [review]

File: tools/slicec-cs/src/generators/class_generator.rs [hunk 4]
```diff
@@ -123,6 +121,10 @@ fn constructor(
 
     let mut builder = FunctionBuilder::new(access, "", escaped_name, FunctionType::BlockBody);
 
+    if fields.iter().any(|f| f.is_required()) || base_fields.iter().any(|f| f.is_required()) {
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply tools/slicec-cs/src/generators/class_generator.rs:124]: This doesn't work, we don't pass all fields to constructor.

File: tools/slicec-cs/src/slicec_ext/member_ext.rs [hunk 5]
```diff
@@ -45,6 +48,12 @@ impl FieldExt for Field {
             .any(|a| a.downcast::<CsReadonly>().is_some())
     }
 
+    fn is_required(&self) -> bool {
+        return !self.data_type().is_optional
+            && !self.data_type().is_value_type()
+            && !matches!(self.parent().concrete_entity(), Entities::Struct(_));
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply tools/slicec-cs/src/slicec_ext/member_ext.rs:54]: See #4023.

**Pull request merged**